### PR TITLE
Add -createUiDefinition and -mainTemplate to managedapp definition create

### PR DESCRIPTION
**Pull request by vivsriaus**
- Add -createUiDefinition and -mainTemplate to managedapp definition create
- Rename appliance client to ApplicationClient

---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

### Command Guidelines

- [x] Each command and parameter has a meaningful description.
- [x] Each new command has a test.

(see [Authoring Command Modules](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules))

### CODE REVIEW

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_params.py [hunk 1]
```diff
@@ -119,16 +119,18 @@
...
 register_cli_argument('managedapp create', 'managedby_resource_group_id', options_list=('--managed-rg-id', '-m'), help='the resource group managed by the managed application')
 register_cli_argument('managedapp create', 'parameters', help='JSON formatted string or a path to a file with such content', type=file_type)
 
-register_cli_argument('managedapp definition create', 'lock_level', **enum_choice_list(ApplianceLockLevel))
+register_cli_argument('managedapp definition create', 'lock_level', **enum_choice_list(ApplicationLockLevel))
 register_cli_argument('managedapp definition create', 'authorizations', options_list=('--authorizations', '-a'), nargs='+', help="space separated authorization pairs in a format of <principalId>:<roleDefinitionId>")
+register_cli_argument('managedapp definition create', 'createUiDefinition', options_list=('--create-ui-def', '-c'), help='JSON formatted string or a path to a file with such content', type=file_type)
+register_cli_argument('managedapp definition create', 'mainTemplate', options_list=('--main-template', '-t'), help='JSON formatted string or a path to a file with such content', type=file_type)
```
Comment (tjprescott) [inline src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_params.py]: Does this actually work?  It should by in python_case, not camelCase. @lmazuel's recent changes to serialization/deserialization might mean that this "just works"...

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py [hunk 2]
```diff
@@ -114,79 +114,98 @@ def create_appliance(resource_group_name,
...
+    :param str application_definition_name:the managed application definition name
     :param str description:the managed application definition description
     :param str display_name:the managed application definition display name
     :param str package_file_uri:the managed application definition package file uri
+    :param str create_ui_def:the managed application definition create ui definition
+    :param str main_template:the managed application definition main template
     :param str tags:tags in 'a=b c' format
     """
+    if(package_file_uri is None and create_ui_def is None and
+            main_template is None):
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py]: maybe simplify like `if not package_file_uri and not create_ui_def and not main_template`? Please apply the same on the couple of conditions below

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py [hunk 3]
```diff
@@ -114,79 +114,98 @@ def create_appliance(resource_group_name,
...
     :param str package_file_uri:the managed application definition package file uri
+    :param str create_ui_definition:the managed application definition create ui definition
+    :param str main_template:the managed application definition main template
     :param str tags:tags in 'a=b c' format
     """
+    if(not package_file_uri and not create_ui_definition and
+       not main_template):
+        raise CLIError('Please specify either --package-file-uri or \
+        --create-ui-definition and --main-template')
+    elif not package_file_uri:
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py]: based on the error message, this should be `elif package_file_uri`

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py [hunk 4]
```diff
@@ -114,79 +114,98 @@ def create_appliance(resource_group_name,
...
+        if not create_ui_definition or not main_template:
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py]: and this should be `if create_ui_definition or main_template`

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py [hunk 5]
```diff
@@ -114,79 +114,98 @@ def create_appliance(resource_group_name,
...
     :param str description:the managed application definition description
     :param str display_name:the managed application definition display name
     :param str package_file_uri:the managed application definition package file uri
+    :param str create_ui_definition:the managed application definition create ui definition
+    :param str main_template:the managed application definition main template
     :param str tags:tags in 'a=b c' format
     """
+    if(not package_file_uri and not create_ui_definition and
+       not main_template):
+        raise CLIError('Please specify either --package-file-uri or \
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py]: please use standard usage error format `usage error:  --package-file-uri PACKAGE_FILE_URI | --create-ui-definition DEFINITION  --main-template TEMPLATE`

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/tests/test_resource.py [hunk 6]
```diff
@@ -569,11 +569,11 @@ def test_show_built_in_policy(self):
 class ManagedAppDefinitionScenarioTest(ScenarioTest):
     @ResourceGroupPreparer()
     def test_managedappdef(self, resource_group):
-        location = 'westcentralus'
+        location = 'eastus2euap'
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/tests/test_resource.py]: Move this to be inside `@ResourceGroupPreparer()`, then expose a resource_group_location in the method signature. Do the same on the new test below

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py [hunk 7]
```diff
@@ -114,79 +114,98 @@ def create_appliance(resource_group_name,
...
-    return racf.appliance_definitions.get(resource_group_name, appliance_definition_name)
+    return racf.application_definitions.get(resource_group_name, application_definition_name)
 
 
-def create_appliancedefinition(resource_group_name,
-                               appliance_definition_name, location,
-                               lock_level, package_file_uri, authorizations,
-                               description, display_name, tags=None):
+def create_applicationdefinition(resource_group_name,
+                                 application_definition_name, location,
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py]: you should default the location to the resource group's location, rather make it manadatory

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py [hunk 8]
```diff
@@ -80,14 +80,14 @@ def create_resource_group(rg_name, location, tags=None):
...
 
 
-def create_appliance(resource_group_name,
-                     appliance_name, managedby_resource_group_id,
-                     location, kind, managedapp_definition_id=None,
-                     plan_name=None, plan_publisher=None, plan_product=None,
-                     plan_version=None, tags=None, parameters=None):
+def create_application(resource_group_name,
+                       application_name, managedby_resource_group_id,
+                       location, kind, managedapp_definition_id=None,
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py]: `location` should be optional with the defaults using resource group's

File: src/command_modules/azure-cli-resource/HISTORY.rst [hunk 9]
```diff
@@ -4,7 +4,9 @@ Release History
 ===============
 (unreleased)
 +++++++++++++++++++
-* policy: support to show built-in policy definition
+* policy: support to show built-in policy definition.
+* managedapp definition: support to create managedapp definition using create-ui-definition and main-template.
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-resource/HISTORY.rst]: please call out `breaking change`

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py (hunk 3 above)
Comment (vivsriaus) [reply src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py]: Thanks

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py (hunk 5 above)
Comment (vivsriaus) [reply src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py]: This command accepts only these combinations:
1 - package file uri
2 - create ui def and main template

So standard usage error format won't apply here

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py (hunk 8 above)
Comment (vivsriaus) [reply src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py]: Done

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py (hunk 7 above)
Comment (vivsriaus) [reply src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py]: Done

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_help.py [hunk 10]
```diff
@@ -38,6 +38,11 @@
             az managedapp definition create -g MyResourceGroup -n MyManagedAppDef -l eastus --display-name "MyManagedAppDef" \\
                 --description "My Managed App Def description" -a "myPrincipalId:myRoleId" --lock-level None \\
                 --package-file-uri "https://path/to/myPackage.zip"
+        - name: Create a managed application defintion with inline values for createUiDefinition and mainTemplate.
+          text: >
+            az managedapp definition create -g MyResourceGroup -n MyManagedAppDef -l eastus --display-name "MyManagedAppDef" \\
+                --description "My Managed App Def description" -a "myPrincipalId:myRoleId" --lock-level None \\
+                --create-ui-definition @myCreateUiDef.json --main-template @myMainTemplate.json
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_help.py]: Should add `min_profile: latest` to this example so it doesn't show on older API versions as I assume it doesn't work on older API versions.

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py (hunk 5 above)
Comment (yugangw-msft) [reply src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py]: the `standard` means the error  starts with `usage error`, and also try to use arg-parser style of error reporting

**Comment by derekbekoe** [comment]
Please verify that this doesn't break the older profile (which is used by Azure Stack).

`az cloud update --profile 2017-03-09-profile` then verify that the commands that shouldn't be available in the older profile do not appear and `--help` works on the remaining commands.

**Comment by vivsriaus** [comment]
@derekbekoe assuming your comment is for my other PR?

**Comment by yugangw-msft** [comment]
@derekbekoe, the managed app is not yet in the stack profile. Policy is there which we have #4478 for that

**Pull request merged**